sql: fix end key selection with tuples

knz opened the pull request:
Prior to this patch, queries using ordering predicates (<, >, <=, >=)
and indices across multiple columns where some columns have no
Prev/Next values (such as STRING, BYTES) would return incorrect
results. This patch corrects this situation.

Thanks to @eisenstatdavid for uncovering this.

Fixes #10412.
Fixes #11673.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10475)
<!-- Reviewable:end -->

knz committed:
sql: fix end key selection with tuples

Prior to this patch, queries using ordering predicates (<, >, <=, >=)
and indices across multiple columns where some columns have no
Prev/Next values (such as STRING, BYTES) would return incorrect
results. This patch corrects this situation.

Thanks to @eisenstatdavid for uncovering this.

This patch also incidentally adds some new functionality
and removes a few bug:

- new feature: unary negation now works with INTERVAL too.

- new feature: input string syntax for intervals now supports
  nanoseconds and microseconds.

- bug fix: the "next" and "previous" operators used internally for
  timestamps was incorrectly using nanosecond precision, whereas only
  microseconds are supported since
  d5a3660955441a2ad1a7af088f4c9ef65272391a.

- bug fix: the upper and lower bounds to compare for float values are
  +/- infinity.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KVpmgFrAkTFN-dD4a3h:b5xs7ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
> ```Go
> 			return false
> 		}
> 		break
> ```

can you help me understand why this change was necessary (and is correct)? better yet, could you write unit tests for this and `(*DTuple).HasNext()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-:-KVpnB4XNLb4EQ41ob-M:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KVqGAFoJRcIzZYwPYHp:b-sfl3pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you help me understand why this change was necessary (and is correct)? better yet, could you write unit tests for this and `(*DTuple).HasNext()`?
</details>
I have added the missing explanatory comment, and in the process of doing so realized that my first implementation was incorrect and that a fix needed slightly more work.

From this point I am unsure how to proceed w.r.t testing. Where would a good test for this live?

@eisenstatdavid could you also suggest good test scenarios?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KVvEbCdhE4TgTOdGP76:b-6pytom) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I have added the missing explanatory comment, and in the process of doing so realized that my first implementation was incorrect and that a fix needed slightly more work.
> 
> From this point I am unsure how to proceed w.r.t testing. Where would a good test for this live?
> 
> @eisenstatdavid could you also suggest good test scenarios?
</details>
I was thinking of quite a narrow unit test in datum_test.go; a table driven test of `Has{Prev,Next}`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Thanks very much for taking a look at this. The code here is subtle due to the unintuitive contract binding `HasNext`, `IsMax`, `Next`. cc @mjibson, who was pondering a simplification at the expense of performance.

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KVzHNeKSoLTLUsh01hR:bn5yh4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I was thinking of quite a narrow unit test in datum_test.go; a table driven test of `Has{Prev,Next}`.
</details>
Soundness here is necessary for correctness; completeness is just for performance. If you wanted to be thorough, you could test `HasNext`, `Next` and friends on all values of a finite tuple domain, performing the betweenness check mentioned in the contract for `Next` whenever you get a value from it.

I would also be happy with a table.

---

*[pkg/sql/parser/datum.go, line 1392 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzJ8zXffCndNScmr-3:-KVzJ8zXffCndNScmr-4:b3pnxwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1392)):*
> ```Go
> 
> // HasPrev implements the Datum interface.
> func (d *DTuple) HasPrev() bool {
> ```

Given that `x.HasPrev` is allowed to return true on the minimum element (see, e.g., DBool), this implementation is still not quite right. If the last element (say) is minimum but is of a type `x` with no maximum, then we break out, `x.HasPrev` in the exit test returns true, and then we end up calling `x.Prev` in `DTuple.Prev` on the minimum. I think this is correct:

```
var i int
for i = len(*d) - 1; i > 0; i-- {
  if !(*d)[i].IsMin() {
    break
  }
  if !(*d)[i].hasMax() {
    return false
  }
}
// If i > 0, then we've already established that (*d)[i] is not min.
// If i == 0, then it doesn't matter if (*d)[i] is min because if so, then the tuple is min.
return (*d)[i].HasPrev()
```

Now that I think about it, though, we need to figure out what to do about `NULL` too. :facepalm: You might decide that it's easier not to do this odometer thing.

---

*[pkg/sql/parser/datum.go, line 1409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzOHfX6NNrMvCZeKJh:-KVzOHfX6NNrMvCZeKJi:brrtkhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1409)):*
> ```Go
> 		break
> 	}
> 	return i >= 0 && (*d)[i].HasPrev()
> ```

The loop guard is `i > 0`, so the `i >= 0 &&` is redundant.

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KVzO8WGSPZYlemLuBEp:b-v3g3r6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
> ```Go
> 	copy(n, *d)
> 	for i := len(n) - 1; i >= 0; i-- {
> 		if n[i].IsMin() && n[i].hasMax() {
> ```

The `hasMax` check here is not required by the contract, but if you want to be safe, it makes sense for the loop body to be

```
if !n[i].IsMin() {
  if !n[i].HasPrev() {
    panic(...)
  }
  n[i] = n[i].Prev()
  return &n
}
if !n[i].hasMax() {
  panic(...)
}
n[i] = n[i].max()
```

---

*[pkg/sql/parser/datum.go, line 1495 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzGq_FIq2s4jAlG_4E:-KVzGq_Gy13cX6Vi46RK:bpp0iwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1495)):*
> ```Go
> 		}
> 	}
> 	return len(*d) > 0
> ```

I'm confused by this test and the other like it. Why can't the empty tuple be both minimum and maximum?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-:-KVzPN4p7KMNuC1wEcpT:b3c07en)*
<!-- Sent from Reviewable.io -->

knz commented:
I made some effort to create a meaningful test suite. Can you have a look at his now and use it to drive the discussion forward?

Also I would be interested in your opinion of the "unintuive contract" -- all my knowledge comes from NormalizeExpr(), and I captured that in the new test, but perhaps you have comments on that now.

---

Review status: 1 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KW323wEFmSAI84wYBQR:b-zfpm5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> Soundness here is necessary for correctness; completeness is just for performance. If you wanted to be thorough, you could test `HasNext`, `Next` and friends on all values of a finite tuple domain, performing the betweenness check mentioned in the contract for `Next` whenever you get a value from it.
> 
> I would also be happy with a table.
</details>
Please have a look at the new table in datum_test.go and tell me what you think of it.

---

*[pkg/sql/parser/datum.go, line 1392 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzJ8zXffCndNScmr-3:-KW32OqJwiawM509oyxu:bjbz3sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1392)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> Given that `x.HasPrev` is allowed to return true on the minimum element (see, e.g., DBool), this implementation is still not quite right. If the last element (say) is minimum but is of a type `x` with no maximum, then we break out, `x.HasPrev` in the exit test returns true, and then we end up calling `x.Prev` in `DTuple.Prev` on the minimum. I think this is correct:
> 
> ```
> var i int
> for i = len(*d) - 1; i > 0; i-- {
>   if !(*d)[i].IsMin() {
>     break
>   }
>   if !(*d)[i].hasMax() {
>     return false
>   }
> }
> // If i > 0, then we've already established that (*d)[i] is not min.
> // If i == 0, then it doesn't matter if (*d)[i] is min because if so, then the tuple is min.
> return (*d)[i].HasPrev()
> ```
> 
> Now that I think about it, though, we need to figure out what to do about `NULL` too. :facepalm: You might decide that it's easier not to do this odometer thing.
</details>
Indeed it's possible that HasPrev returns true on a tuple that has no previous value, however I'm rather confident in that case that IsMin() is true too so Prev() will not be called.

The contract if I understand correctly is that Prev() can only be called if HasPrev() is true and IsMin() is false, and both need to be checked first.
If my understanding is wrong please let me know.

---

*[pkg/sql/parser/datum.go, line 1409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzOHfX6NNrMvCZeKJh:-KW31fir_kQozovl9zXz:b-vupd9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1409)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> The loop guard is `i > 0`, so the `i >= 0 &&` is redundant.
</details>
In Go like in C the increment (i--) is computed before the guard so when the loop ends without break, i < 0 (in which case there is no prev).

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KW3AxUdCCpiIAJqR9H6:bypr9c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> The `hasMax` check here is not required by the contract, but if you want to be safe, it makes sense for the loop body to be
> 
> ```
> if !n[i].IsMin() {
>   if !n[i].HasPrev() {
>     panic(...)
>   }
>   n[i] = n[i].Prev()
>   return &n
> }
> if !n[i].hasMax() {
>   panic(...)
> }
> n[i] = n[i].max()
> ```
</details>
I'm not sure this is equivalent. How did you come to it?

---

*[pkg/sql/parser/datum.go, line 1495 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzGq_FIq2s4jAlG_4E:-KW39HkRw8BXJMiWyExq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1495)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> I'm confused by this test and the other like it. Why can't the empty tuple be both minimum and maximum?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-:-KW3AVstJiIzZd9PZlsh:bx43c19)*
<!-- Sent from Reviewable.io -->

knz commented:
In particular now I've worked on this some more I noticed a couple of strange areas where perhaps additional attention is needed:

- what's up with min/max on floats? why wasn't it -/+ inf before? @nvanbenschoten?
- am I right to understand there was a bug with prev/next on timestamps?
- any suggestions as to min/max on dates and timestamps?

knz commented:
Also @RaduBerinde and @mjibson if you could have a general look at the direction of this PR and the various example/test cases in the new `TestDatumOrdering` that would be great.

eisenstatdavid commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/datum.go, line 1157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVpmgFrAkTFN-dD4a3g:-KW3NRNhsZEDxiDED3av:b-64ui9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfd85c8cf30ae228a627d1075848c3dea63d1a9a/pkg/sql/parser/datum.go#L1157)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Please have a look at the new table in datum_test.go and tell me what you think of it.
</details>
Looks good, comment below.

---

*[pkg/sql/parser/datum.go, line 1392 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzJ8zXffCndNScmr-3:-KW3NYBZ8zf74zEyblFx:bdimhi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1392)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Indeed it's possible that HasPrev returns true on a tuple that has no previous value, however I'm rather confident in that case that IsMin() is true too so Prev() will not be called.
> 
> The contract if I understand correctly is that Prev() can only be called if HasPrev() is true and IsMin() is false, and both need to be checked first.
> If my understanding is wrong please let me know.
</details>
Yes, this is the right contract.

---

*[pkg/sql/parser/datum.go, line 1409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzOHfX6NNrMvCZeKJh:-KW3GmRneAW0DoBn9Zhb:bbpnmq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1409)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> In Go like in C the increment (i--) is computed before the guard so when the loop ends without break, i < 0 (in which case there is no prev).
</details>
The loop condition is `i > 0`, so the decrement never proceeds when `i == 0`. The actual problem here is the empty tuple, which initializes `i = -1`. I revise my suggestion to

```
for i := len(*d) - 1; i >= 0; i-- {
  if !(*d)[i].IsMin() {
    return (*d)[i].HasPrev()
  }
  if !(*d)[i].hasMax() {
    return false
  }
}
return false
```

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KW3II4fgkOT1_FNbh0P:broxkqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'm not sure this is equivalent. How did you come to it?
</details>
By fixing the bug here and before where a hypothetical natural number type like

```
type DNatural int // Let's pretend that this is a bignum type
func (d *DNatural) IsMin() bool { return *d == 0 }
func (*DNatural) HasPrev() bool { return true }
func (d *DNatural) Prev() Datum {
  if *d <= 0 { panic("failed precondition") }
  p := *d - 1
  return &p
}
func (*DNatural) hasMax() bool { return false }
func (*DNatural) max() Datum { panic("failed precondition") }
```

causes `(1:natural, 0:natural).HasPrev()` to be true, and `(1:natural, 0:natural).Prev()` to panic (legit, since `(1, 0)` is not minimum).

---

*[pkg/sql/parser/datum_test.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KW3LcDHOtiCeeanzFkz:-KW3LcDHOtiCeeanzFl-:bd4cd1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e77e315b96f91a32e827cf24854b84be6a83976/pkg/sql/parser/datum_test.go#L148)):*
> ```Go
> 			`(-9223372036854775808, NULL)`, `(9223372036854775807, NULL)`},
> 		{`row(NULL, 10)`, `(NULL, 9)`, `(NULL, 11)`,
> 			`(NULL, -9223372036854775808)`, `(NULL, 9223372036854775807)`},
> ```

Looks reasonable. I'd toss in a longer tuple or two with a bigger ripple on `Prev/Next`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@knz I'm not sure why Floats didn't just use +/- inf before, but based on the comment (`// Using >= accounts for +inf as well.`), it seems deliberate. The commit that introduced that line was 6c50d753c89990079aa5e35b2d0ce79d41cdda6b. Postgres has a notion of +/- Infinity as well for floating point numbers, so this doesn't seem to be a decision made for compatibility reasons. Perhaps @petermattis remembers a deeper rationale for the choice.

knz commented:
@nvanbenschoten I needed a couple of "minor" array-related tweaks to create the tests needed in the final comment. Please review these too :)

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/datum.go, line 1409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzOHfX6NNrMvCZeKJh:-KXh5NGmnQSlw9nWHKZi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1409)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

The loop condition is `i > 0`, so the decrement never proceeds when `i == 0`. The actual problem here is the empty tuple, which initializes `i = -1`. I revise my suggestion to

```
for i := len(*d) - 1; i >= 0; i-- {
  if !(*d)[i].IsMin() {
    return (*d)[i].HasPrev()
  }
  if !(*d)[i].hasMax() {
    return false
  }
}
return false
```
</blockquote></details>

Done.

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KXh7UtHHRiw6PtkLKpL:bjqkphg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

By fixing the bug here and before where a hypothetical natural number type like

```
type DNatural int // Let's pretend that this is a bignum type
func (d *DNatural) IsMin() bool { return *d == 0 }
func (*DNatural) HasPrev() bool { return true }
func (d *DNatural) Prev() Datum {
  if *d <= 0 { panic("failed precondition") }
  p := *d - 1
  return &p
}
func (*DNatural) hasMax() bool { return false }
func (*DNatural) max() Datum { panic("failed precondition") }
```

causes `(1:natural, 0:natural).HasPrev()` to be true, and `(1:natural, 0:natural).Prev()` to panic (legit, since `(1, 0)` is not minimum).
</blockquote></details>

Thanks for the example. Although in this particular case again the contract would prevent the panic, because one cannot call Prev() without first checking both HasPrev() and IsMin(); in this case IsMin() is true.

Now if you look at the code I wrote and your revised suggestion for HasPrev() above (which I have accepted) you will see that the conditions I used here now properly mirror the conditions you propose above. It was my check on HasPrev here which I think was redundant with the contract. PTAL.

---

*[pkg/sql/parser/datum_test.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KW3LcDHOtiCeeanzFkz:-KXh8UBcW_axxN_k6UpM:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e77e315b96f91a32e827cf24854b84be6a83976/pkg/sql/parser/datum_test.go#L148)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Looks reasonable. I'd toss in a longer tuple or two with a bigger ripple on `Prev/Next`.
</blockquote></details>

Good call. Done.

---

*[pkg/sql/parser/datum_test.go, line 165 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXhTDGsmXVbtuhRbkmc:-KXhTDGt09qQH_DjjE0v:b-u115vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2d4b2eae2b0bbf6aecc2e2a3d737db2b59379f/pkg/sql/parser/datum_test.go#L165)):*
> ```Go
> 			`((false), (false))`, `((true), (true))`},
> 
> 		// Arrays
> ```

Also added mixed array/tuple types. PTAL.
(ping @nvanbenschoten too)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-:-KXhSkuxvAl6CEG78sop:b-bhufir)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 6 files at r7.
Review status: 1 of 10 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 85 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlAHwzUkvmupRHDIL2:-KXlAHwzUkvmupRHDIL3:b-xzpram) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L85)):*
> ```Go
> 	IsMin() bool
> 
> 	// max() returns the upper value, if one exists. Only defined if
> ```

I dont think the expansion of this interface is warranted because we now have three methods that are doing very similar things and duplicate information. Instead, couldn't we have a single method in the interface:
```
// returns the min Datum value and a flag stating if the min value exists.
Min() (Datum, ok)
```
And a related function
```
func IsMin(d Datum) bool {
  if min, ok := d.Min(); ok {
    return d.Compare(min) == 0
  }
  return false
}
```
Are there any benefits to the current approach that I'm missing?

---

*[pkg/sql/parser/datum.go, line 1919 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlCMIlB8EzrCxU11_5:-KXlCMIlB8EzrCxU11_6:bxn4nqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L1919)):*
> ```Go
> 
> // Next implements the Datum interface.
> func (d *DArray) Next() Datum {
> ```

I thought we had decided that arrays are variable-length, so the Next behavior should be closer to a byte string than to a tuple (ie. Next appends the minimum value of the parameterized-type).

---

*[pkg/sql/parser/datum_test.go, line 225 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlEcNgc8jnpFent4vb:-KXlEcNgc8jnpFent4vc:blfj5hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum_test.go#L225)):*
> ```Go
> 			}
> 		} else {
> 			if !d.HasPrev() && td.prev != "min" {
> ```

Could we pull `"min"` and `"max"` into constants?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: 1 of 10 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KXr-pY7t3Qh3koY3j09:bhoymc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Thanks for the example. Although in this particular case again the contract would prevent the panic, because one cannot call Prev() without first checking both HasPrev() and IsMin(); in this case IsMin() is true.

Now if you look at the code I wrote and your revised suggestion for HasPrev() above (which I have accepted) you will see that the conditions I used here now properly mirror the conditions you propose above. It was my check on HasPrev here which I think was redundant with the contract. PTAL.

</blockquote></details>

My suggestion was if you want a violation of the tuple Datum's contract (calling `Prev()` on an invalid argument) not to potentially cause a violation of the element Datum's contract. If you don't care about this property, then

```
if !n[i].IsMin() {
  n[i] = n[i].Prev()
  break
}
n[i] = n[i].max()
```

avoids the `hasMax` call -- `DTuple.HasPrev` has already ensured that, when `n[i]` is minimum, it has a max.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 10 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 1418 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KVzO8WFhN3nhwNpxgm7:-KYE9wYBYF18CO4YD9xS:bw74jo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9567afdb6fa05aaf5c9483d1db12272ed33b4a72/pkg/sql/parser/datum.go#L1418)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

My suggestion was if you want a violation of the tuple Datum's contract (calling `Prev()` on an invalid argument) not to potentially cause a violation of the element Datum's contract. If you don't care about this property, then

```
if !n[i].IsMin() {
  n[i] = n[i].Prev()
  break
}
n[i] = n[i].max()
```

avoids the `hasMax` call -- `DTuple.HasPrev` has already ensured that, when `n[i]` is minimum, it has a max.
</blockquote></details>

Good suggestion. Thanks for explaining. Done.

---

*[pkg/sql/parser/datum.go, line 85 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlAHwzUkvmupRHDIL2:-KYE7e869YqicEDfncA9:b82scx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L85)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I dont think the expansion of this interface is warranted because we now have three methods that are doing very similar things and duplicate information. Instead, couldn't we have a single method in the interface:
```
// returns the min Datum value and a flag stating if the min value exists.
Min() (Datum, ok)
```
And a related function
```
func IsMin(d Datum) bool {
  if min, ok := d.Min(); ok {
    return d.Compare(min) == 0
  }
  return false
}
```
Are there any benefits to the current approach that I'm missing?
</blockquote></details>

We can't merge `hasMin`/`hasMax` with `min`/`max` as long as `HasPrev`/`HasNext` are separated from `Prev`/`Next` (we wouldn't want to compute the minimum in `HasPrev`, discard it immediately, then compute it again in `Prev`). I'll see if I can merge HasPrev/Prev in one method separately. Thanks for the hint.

---

*[pkg/sql/parser/datum.go, line 1919 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlCMIlB8EzrCxU11_5:-KYECP7s3Tw4_IYADrg9:brraagv) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L1919)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I thought we had decided that arrays are variable-length, so the Next behavior should be closer to a byte string than to a tuple (ie. Next appends the minimum value of the parameterized-type).
</blockquote></details>

Aha, thanks. Since NULL sorts before all other values, I assume that the extra item should be NULL rather than the minimum of the parameterized type. What do you think?

---

*[pkg/sql/parser/datum_test.go, line 225 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlEcNgc8jnpFent4vb:-KYECavMm0rO8u2jcWKx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum_test.go#L225)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could we pull `"min"` and `"max"` into constants?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

knz committed:
sql/parser: simplify the Datum interface.

Prior to this patch Datum had a rather complex interface with separate
HasPrev/Prev/hasMin/min methods (and HasNext/Next/hasMax/max). This
patch simplifies by merging HasPrev/Prev, HastNext/Next, hasMin/min
and hasMax/max.

knz commented:
I have added an extra commit at the end to remove the redundancy in the interfaces. PTAL!

eisenstatdavid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I don't want to gate this PR on fixing our null handling -- see below.

---

Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/index_selection.go, line 655 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KYEfSpfY8FX03XSusSk:-KYEfSpgWbwohwwtvxaS:b-o0yhoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/289975de6d0b2496aae83a1c2e52aac0d5d6cc47/pkg/sql/index_selection.go#L655)):*
> ```Go
> 						)
> 					} else {
> 						if nextRightVal, hasNext := c.Right.(parser.Datum).Next(); hasNext {
> ```

Can we collapse to an else if?

---

*[pkg/sql/parser/datum.go, line 1919 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlCMIlB8EzrCxU11_5:-KYEiWXXZlaGK5dlCksh:bhl9cri) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L1919)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Aha, thanks. Since NULL sorts before all other values, I assume that the extra item should be NULL rather than the minimum of the parameterized type. What do you think?
</blockquote></details>

Probably null, but there's a mess here.

The logic for not defining the previous element of (e.g.) -inf to be null is that, in a filter expression like col < -inf, the null values of col don't satisfy the filter (because null < -inf is null (unknown, technically, but we conflate unknown and boolean null) when we're using the user-accessible builtin and not .Compare), and thus should be excluded.

For arrays, I think Next(a) = a + [null] is good enough for the limited purpose of scanning an index. If we unthinkingly rewrite y > [1] to y >= [1, null], then despite the fact that [1, 2] > [1] is true, [1, 2] > [1, null] should be unknown (guessing this is what PostgreSQL does from its handling of tuples; IIRC, we don't do this because the array builtin calls .Compare, but we should).

For tuples also, we don't have a good answer. (2, null) sorts in between (1, max) and (2, min), so we should include (2, null) when there's a > (1, max) filter. This means that defining Next((1, max)) = (2, min) is bad, because then it's not safe to rewrite, e.g., y > (1, max) to y >= (2, min), because the semantics for (2, null) changes, from true to either false (.Compare) or null (as the builtin should return, but doesn't yet). Defining Next((1, max)) = (2, null) has the problem described above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-:-KYEwBd0oLiALv8UCGTL:b-xm6ule)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/index_selection.go, line 655 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KYEfSpfY8FX03XSusSk:-KYFCu121usNZ3UjpSh7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/289975de6d0b2496aae83a1c2e52aac0d5d6cc47/pkg/sql/index_selection.go#L655)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Can we collapse to an else if?
</blockquote></details>

Done.

---

*[pkg/sql/parser/datum.go, line 1919 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475#-KXlCMIlB8EzrCxU11_5:-KYFCuhAI_LSJtpfKNu0:b-3rqjzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/40739a59632aeb894eba74999c3fdb592c509b8b/pkg/sql/parser/datum.go#L1919)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary><blockquote>

Probably null, but there's a mess here.

The logic for not defining the previous element of (e.g.) -inf to be null is that, in a filter expression like col < -inf, the null values of col don't satisfy the filter (because null < -inf is null (unknown, technically, but we conflate unknown and boolean null) when we're using the user-accessible builtin and not .Compare), and thus should be excluded.

For arrays, I think Next(a) = a + [null] is good enough for the limited purpose of scanning an index. If we unthinkingly rewrite y > [1] to y >= [1, null], then despite the fact that [1, 2] > [1] is true, [1, 2] > [1, null] should be unknown (guessing this is what PostgreSQL does from its handling of tuples; IIRC, we don't do this because the array builtin calls .Compare, but we should).

For tuples also, we don't have a good answer. (2, null) sorts in between (1, max) and (2, min), so we should include (2, null) when there's a > (1, max) filter. This means that defining Next((1, max)) = (2, min) is bad, because then it's not safe to rewrite, e.g., y > (1, max) to y >= (2, min), because the semantics for (2, null) changes, from true to either false (.Compare) or null (as the builtin should return, but doesn't yet). Defining Next((1, max)) = (2, null) has the problem described above.
</blockquote></details>

Wow yes thanks this is very broken. Filed separately as #12022.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10475)*
<!-- Sent from Reviewable.io -->

knz commented:
Merge pending on #11703.
